qemu/tests/migration_with_numa.py

```diff
@@ -0,0 +1,104 @@
+        return nodes_size
+
+    host_nodes_size = get_nodes_size(size_type='MemFree')
+    mem_devs = params.objects('mem_devs')
+    num = len(mem_devs)
+    for mem_dev in mem_devs:
+        size_mem = params.object_params(mem_dev).get('size_mem')
+        size_mem = float(normalize_data_size(size_mem, 'K'))
+        for node, size in host_nodes_size.items():
+            if int(size) >= int(size_mem):
+                params['host-nodes_mem_%s' % mem_dev] = str(node)
+                num -= 1
+                del host_nodes_size[node]
+                break
+
+    if num != 0:
+        test.cancel("No enough available nodes for testing!")
+
+    params['start_vm'] = 'yes'
+    env_process.preprocess(test, params, env)
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    # do migration
+    mig_timeout = float(params.get("mig_timeout", "3600"))
+    mig_protocol = params.get("migration_protocol", "tcp")
+    vm.migrate(mig_timeout, mig_protocol, env=env)
+    session = vm.wait_for_login()
+
+    os_type = params["os_type"]
+    if os_type == 'linux':
+        error_context.context("Check the memory in guest", logging.info)
+        # Use 30 plus the gap of 'MemTotal' in OS and '-m' in cli as threshold
+        MemTotal = get_mem_info(session, 'MemTotal')
+        MemTotal = float(normalize_data_size('%s KB' % MemTotal))
+        error_context.context("MemTotal in guest os is %s MB"
+                              % MemTotal, logging.info)
+        threshold = float(params.get_numeric("mem") - MemTotal) + 30
+        error_context.context("The acceptable threshold is: %s"
+                              % threshold, logging.info)
+        guest_nodes_size = get_nodes_size(size_type='MemTotal', session=session)
+        for nodenr, node in enumerate(params.objects('guest_numa_nodes')):
+            mdev = params.get("numa_memdev_node%d" % nodenr)
+            if mdev:
+                mdev = mdev.split('-')[1]
+                size = float(normalize_data_size(params.get("size_mem_%s" % mdev)))
+                if size - guest_nodes_size[nodenr] > threshold:
+                    test.fail("[Guest]Wrong size of numa node %d: %f. Expected:"
+                              " %s" % (nodenr, guest_nodes_size[nodenr], size))
+
+    error_context.context("Check the memory in dest host", logging.info)
+    qemu_pid = vm.get_pid()
+    for mem_dev in mem_devs:
+        memdev_params = params.object_params(mem_dev)
+        size_mem = memdev_params.get('size_mem')
+        size_mem = float(normalize_data_size(size_mem, 'K'))
+        smaps = process.getoutput("grep -E -B1 '^Size: *%d' /proc/%d/smaps"
+                                  % (size_mem, qemu_pid))
+        mem_start = smaps.split('\n')[0].split('-')[0]
+        numa_maps = process.getoutput("grep %s /proc/%d/numa_maps"
+                                      % (mem_start, qemu_pid))
+        bind_node = re.search(r'bind:(\d+)', numa_maps).groups()[0]
```

Comment:
It seemed there's something wrong on ppc, can you re-check it ?  By the way, the numbers  of numa nodes  are not consecutive on the host,  does it also cover this situation ? Thanks.
AttributeError: 'NoneType' object has no attribute 'groups'

Reply:
@MiriamDeng  As we discussed, this case find the host node with suitable memory size, and use the corresponding nodeid, so it does not matter whether the numa nodes are consecutive or not.

For the `NoneType` attribute error, i can't reproduce it with my p9 system. Let's confirm it with more system to check if it is host related, and wait for your latest result, thanks!